[wallet] Use destination groups instead of coins in coin select

kallewoof opened the pull request:
This PR adds an optional (off by default) `-avoidpartialspends` flag, which changes coin select to use output groups rather than outputs, where each output group corresponds to all outputs with the same destination.

It is a privacy improvement, as each time you spend some output, any other output that is publicly associated with the destination (address) will also be spent at the same time, at the cost of fee increase for cases where coin select without group restriction would find a more optimal set of coins (see example below).

For regular use without address reuse, this PR should have no effect on the user experience whatsoever; it only affects users who, for some reason, have multiple outputs with the same destination (i.e. address reuse).

Nodes with this turned off will still try to avoid partial spending, if the fee of the resulting transaction is not greater than the fee of the original transaction.

Example: a node has four outputs linked to two addresses `A` and `B`:

* 1.0 btc to `A`
* 0.5 btc to `A`
* 1.0 btc to `B`
* 0.5 btc to `B`

The node sends 0.2 btc to `C`. Without `-avoidpartialspends`, the following coin selection will occur:
* 0.5 btc to `A` or `B` is picked
* 0.2 btc is output to `C`
* 0.3 - fee is output to (unique change address)

With `-avoidpartialspends`, the following will instead happen:
* Both of (0.5, 1.0) btc to `A` or `B` is picked (one or the other pair)
* 0.2 btc is output to `C`
* 1.3 - fee is output to (unique change address)

As noted, the pro here is that, assuming nobody sends to the address after you spend from it, you will only ever use one address once. The con is that the transaction becomes slightly larger in this case, because it is overpicking outputs to adhere to the no partial spending rule.

This complements #10386, in particular it addresses @luke-jr and @gmaxwell's concerns in https://github.com/bitcoin/bitcoin/pull/10386#issuecomment-300667926 and https://github.com/bitcoin/bitcoin/pull/10386#issuecomment-302361381.

Together with `-avoidreuse`, this fully addresses the concerns in #10065 I believe.

kallewoof committed:
test: Add basic testing for wallet groups

luke-jr commented:
It doesn't avoid reuse (you still need to sign for each input), and has security issues in the event of QC (but that's already a hypothetical big issue for us anyway), but seems useful anyway.

Is this affected by the current coin control grouping defect, where it includes change in the same group as unrelated outputs? That would be a serious privacy risk, I think, since it would basically tell the world which output was change.

kallewoof commented:
@luke-jr:
> It doesn't avoid reuse (you still need to sign for each input), and has security issues in the event of QC (but that's already a hypothetical big issue for us anyway), but seems useful anyway.

Together with `-avoidreuse` in #10386 it avoids reuse, I think. Are there any cases I'm missing?

> Is this affected by the current coin control grouping defect, where it includes change in the same group as unrelated outputs? That would be a serious privacy risk, I think, since it would basically tell the world which output was change.

I'm not sure what this defect is. Is it described somewhere?

kallewoof commented on src/wallet/coincontrol.h:
```diff
@@ -27,6 +27,8 @@ class CCoinControl
     //! Override the default confirmation target if set
     boost::optional<unsigned int> m_confirm_target;
+    //! Avoid partial use of funds sent to a given address: -1 means default
+    short m_avoid_partial_spends;
```
I want to avoid using a `bool` here because it's important to distinguish `true`/`false` from 'unset', which should use `DEFAULT_AVOIDPARTIALSPENDS`.

ziggamon commented:
Does this also keep track of the change outputs in case more coins arrive later to `A`?

Imagine your example, where the coins from `A` go to `C`, with a change output to what we now call `D`.

If more coins arrive at `A` later, will they be spent together with `D`?

kallewoof commented:
@ziggamon Change outputs will go to a different destination so they will be considered separate from their inputs. In other words, `D` in your case will not be considered related to `A` in this case, and will not be associated with any coins arriving at `A` afterwards.

NicolasDorier commented on src/wallet/wallet.cpp:
```diff
@@ -3063,6 +3048,30 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
     return true;
 }
 
+bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletTx& wtxNew, CReserveKey& reservekey, CAmount& nFeeRet,
```
I think this is avoidable complexitiy, because `avoidpartial` will be always higher fees.
I think it is OK, you pay fee for privacy, while giving incentive to privacy over fee by default.

kallewoof replied:
Actually `avoidpartial` gives the same fee a lot of the times, from what I've seen so far. It depends on your utxo set, of course.

NicolasDorier commented on src/wallet/wallet.cpp:
```diff
@@ -2575,17 +2556,21 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
 
+    // form groups from remaining coins; note that preset coins will not
+    // automatically have their associated (same address) coins included
+    std::vector<OutputGroup> groups = group_outputs(vCoins, !(coinControl && coinControl->m_avoid_partial_spends > -1 ? coinControl->m_avoid_partial_spends : gArgs.GetBoolArg("-avoidpartialspends", DEFAULT_AVOIDPARTIALSPENDS)));
```
I think `coinControl->m_avoid_partial_spends` should be a boolean, and `coinControl` should set its value based on the arg `avoidpartialspends` just after being created. (At worse, just before this call)

This will make the code easier to reason about.

kallewoof replied:
Unfortunately that means either adding a coincontrol.cpp file and moving the constructor there, or adding a dependency on the `gArg` thing in `coincontrol.h`, which seems a bit overkill just for this. I'm very open to suggestions though; the short thing is an undesired hack

NicolasDorier commented on src/wallet/coincontrol.h:
```diff
@@ -30,6 +30,8 @@ class CCoinControl
     //! Override the default confirmation target if set
     boost::optional<unsigned int> m_confirm_target;
+    //! Avoid partial use of funds sent to a given address: -1 means default
+    short m_avoid_partial_spends;
```
what about `std::optional<bool>` ?

MarcoFalke replied:
(since C++17) 😛

NicolasDorier replied:
Ah I forgot we were using boost for that... I would still prefer boost than a short though.

kallewoof replied:
Changed to using a `boost::optional`.

kallewoof commented:
@NicolasDorier 
> Making it default is the right way, but we should keep in mind that people relying on address reuse will pay way higher fee, so they need to be notified appropriately.

I think keeping it off by default for now is reasonable, until we get enough of an idea of the general consensus among people. I suspect, though, that most people won't care, as long as it doesn't give overly huge fee increases.

> Virtually all japanese exchanges are using address reuse (1 address per customer), this change would impact them a lot. I am unsure if they depend on Bitcoin Core Wallet though.

Also some people have a "permanent" address that they receive payments to over a long period of time (e.g. charities). They also care little about privacy, usually, as their address is publicly tied to them anyway.

> I am unsure if Bitcoin Core filter uneconomical outputs (filtering outputs which are more costly than their value to spend), but if bitcoin core implements this, we need to make sure this rule is always enforced at the Output level and not at the OutputGroup level. (Need to be enforced before you create the group)

Good point. I think the dust limit check can be added to the grouping method fairly easily.

NicolasDorier commented on src/wallet/wallet.cpp:
```diff
@@ -2421,103 +2421,84 @@ static void ApproximateBestSubset(const std::vector<CInputCoin>& vValue, const C
     }
 }
 
-bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const int nConfMine, const int nConfTheirs, const uint64_t nMaxAncestors, std::vector<COutput> vCoins,
```
I am unsure yet, but I think you can make the diff way smaller by keeping the old signature of this method and building the `OutputGroup` at the scope of this function only.

kallewoof replied:
This method doesn't know whether you want single or grouped as it doesn't have coincontrol. Would have to add a bool for that at least. Not sure it'll become smaller though. It'll just move code around.

NicolasDorier replied:
So the way I did it with NBitcoin is to keep the concept of group internal to only the method which does coin packing.

If grouping is deactivated, I group by random, that way there is a single code path for everything.

I think it is a way smaller code change because all consumer of `SelectCoinsMinConf` do not have to change (you can make the `bool` default to `false` to keep original behavior)
The concept of group does not make lot's of sense anywhere outside of this method.

kallewoof replied:
I discussed with @NicolasDorier and we concluded that the current way this is done is okay (NBitcoin does it differently so it made sense there).

NicolasDorier commented on doc/man/bitcoind.1:
```diff
@@ -389,6 +389,14 @@ Include IP addresses in debug output (default: 0)
+\fB\-avoidpartialspends\fR
+.IP
+Group outputs by address, selecting all or none, instead of selecting on
+a per\-output basis. Improved privacy as an address is only used
```
would say `Privacy is improved as an address...`

kallewoof replied:
Changed

luke-jr commented on src/init.cpp:
```diff
@@ -462,6 +462,8 @@ std::string HelpMessage(HelpMessageMode mode)
         strUsage += HelpMessageOpt("-maxsigcachesize=<n>", strprintf("Limit sum of signature cache and script execution cache sizes to <n> MiB (default: %u)", DEFAULT_MAX_SIG_CACHE_SIZE));
         strUsage += HelpMessageOpt("-maxtipage=<n>", strprintf("Maximum tip age in seconds to consider node in initial block download (default: %u)", DEFAULT_MAX_TIP_AGE));
     }
+    strUsage += HelpMessageOpt("-avoidpartialspends", strprintf(_("Group outputs by address, selecting all or none, instead of selecting on a per-output basis. Privacy is improved as an address is only used once (unless someone sends to it "
```
"by key". Outputs don't have addresses.

kallewoof replied:
I see what you're saying, but in this case I think 'address' is fine; we speak of address reuse, after all. While outputs themselves don't have addresses, the funding to each output was done to a specific address; if that address was the same, then this PR will group them together.

Maybe `Group outputs by key (address), selecting all or...` would be fine, though?

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -47,6 +47,8 @@ OutputType g_change_type = OUTPUT_TYPE_NONE;
 const char * DEFAULT_WALLET_DAT = "wallet.dat";
 const uint32_t BIP32_HARDENED_KEY_LIMIT = 0x80000000;
 
+#define COINCONTROL_AVOID_PARTIAL_SPENDS(cc) ((cc) && (cc)->m_avoid_partial_spends ? (cc)->m_avoid_partial_spends.get() : gArgs.GetBoolArg("-avoidpartialspends", DEFAULT_AVOIDPARTIALSPENDS))
```
Seems this would be better as a method on CCoinControl...

kallewoof replied:
Coin control needs `gArgs` in that case, and currently there is no `.cpp` file so it'd have to be included in the header.

I'm actually starting to think maybe coin control could use a `.cpp` file...

kallewoof commented:
I added `coincontrol.cpp` and moved `SetNull` into it, setting it to the default given in `DEFAULT_AVOIDPARTIALSPENDS`. This cleans up use as it is always a bool with an up-to-date value even when default is used.

luke-jr commented on src/validation.h:
```diff
@@ -53,6 +53,8 @@ static const bool DEFAULT_WHITELISTFORCERELAY = true;
 //! -maxtxfee default
 static const CAmount DEFAULT_TRANSACTION_MAXFEE = 0.1 * COIN;
+//! -avoidpartialspends default
+static const bool DEFAULT_AVOIDPARTIALSPENDS = false;
```
Similarly, in `wallet/wallet.h`

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2564,17 +2545,21 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, std::min((size_t)4, nMaxChainLength/3), groups, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, nMaxChainLength/2, groups, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, nMaxChainLength, groups, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && !fRejectLongChains && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, std::numeric_limits<int64_t>::max(), groups, setCoinsRet, nValueRet));
```
Change from uint64_t max to int64_t max should be documented somewhere, probably done alone in a separate bugfix commit.

kallewoof replied:
Split into its own commit, and documented change in commit message.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2410,103 +2410,84 @@ static void ApproximateBestSubset(const std::vector<CInputCoin>& vValue, const C
+    // we prioritize the lowest larger if
+    // 1. it has multiple outputs (i.e. it is the result of address reuse)
+    // 2. it is not substantially (>2x) larger than the target value
+    if (nTotalLower < nTargetValue || (lowest_larger && lowest_larger->m_outputs.size() > 1 && lowest_larger->m_value < 2 * nTargetValue)) {
```
This policy change should be documented in the commit message.

kallewoof replied:
Split into its own commit, and documented policy change.

kallewoof commented on src/init.cpp:
```diff
@@ -462,6 +462,8 @@ std::string HelpMessage(HelpMessageMode mode)
         strUsage += HelpMessageOpt("-maxsigcachesize=<n>", strprintf("Limit sum of signature cache and script execution cache sizes to <n> MiB (default: %u)", DEFAULT_MAX_SIG_CACHE_SIZE));
         strUsage += HelpMessageOpt("-maxtipage=<n>", strprintf("Maximum tip age in seconds to consider node in initial block download (default: %u)", DEFAULT_MAX_TIP_AGE));
     }
+    strUsage += HelpMessageOpt("-avoidpartialspends", strprintf(_("Group outputs by address, selecting all or none, instead of selecting on a per-output basis. Privacy is improved as an address is only used once (unless someone sends to it "
```
Moved

laanwj commented:
Concept ACK

NicolasDorier commented on src/util.h:
```diff
@@ -357,4 +358,14 @@ std::unique_ptr<T> MakeUnique(Args&&... args)
 
+//! Simplification of std insertion
+template <typename Tdst, typename Tsrc>
+void insert(Tdst& dst, const Tsrc&& src) {
```
As we discussed you might not need `&&` because you can pass temporary object to the function with `const Tsrc&`

NicolasDorier commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,10 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
     }
 }
 
```
nit: You don't have to do this, but I think it would be better if this commit was self contained, as now the clients of `TransactionWithinChainLimit` are not compiling anymore. To very your change is valid we need to check other commits later. This make it harder to review.

NicolasDorier commented:
Code Review ACK, though it would be nice to have the remaining TODO in the test done.

kallewoof commented:
> I am also wondering potential performance problem for wallet with lots of UTXOs. I already heard complaints that core was getting quite slow at doing transactions when for big wallets. (though this could be worked around by asking people to turn on avoidpartialspends)

I think wallets with a ton of outputs would want to use avoidpartialspends as it would affect them minimally fee-wise, and it would improve privacy for odd-ball reuse.

kallewoof committed:
wallet: Remove deprecated OutputEligibleForSpending

kallewoof commented:
I rebased this on top of the new coin selection algorithm.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4387,7 +4387,7 @@ std::vector<COutput>::iterator OutputGroup::discard(const COutput& output) {
     std::vector<CInputCoin> coins;
     for (const auto& output : m_outputs) {
-        coins.emplace_back(output.tx, output.i);
+        coins.emplace_back(output.tx->tx, output.i);
```
Shouldn't this change be part of 27872fa1cf4f7ca5e0c136554078448dc7f92d1d?

kallewoof replied:
Yes! Thanks for noticing.

achow101 reviewed (commented):
Light utACK. I kind of scanned over the functional tests.

MarcoFalke commented:
Needs rebase if still relevant

kallewoof commented:
Rebased.

kallewoof commented:
Rebased.

kallewoof commented:
Rebased

gmaxwell commented:
ACK

sipa commented on src/wallet/wallet.h:
```diff
@@ -669,6 +668,39 @@ struct CoinEligibilityFilter
+    CAmount long_term_fee;
+
+    OutputGroup() : OutputGroup(std::vector<COutput>(), true, 0, 999, 0, 0) {}
+    OutputGroup(const std::vector<COutput>&& outputs, bool from_me, CAmount value, int depth, size_t ancestors, size_t descendants)
```
Const rvalue references aren't very useful.

sipa commented on src/wallet/coinselection.cpp:
```diff
@@ -3,12 +3,13 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <wallet/coinselection.h>
+#include <wallet/wallet.h>
```
Can you avoid this cyclic module dependency (between coinselection and wallet) by moving `OutputGroup` to coinselection.h?

kallewoof replied:
`OutputGroup` depends heavily on `COutput`, which is integral to the wallet. I could move it, but it feels like `COutput` doesn't really belong in coin selection.

sipa replied:
Why not? See the wallet as a user of the coin selection algorithm, filling in the relevant data in its native data type.

sipa replied:
Oh, I see. COutput contains a CWalletTx. Yeah, no, that doesn't belong in coinselection.

But why does the coin selection code need access to CWalletTx's?

kallewoof replied:
It needs the `tx->tx->vout[output.i].nValue` and `tx->IsEquivalentTo()` (but can probably work around the latter).

I could replace this with a new struct without the `CWalletTx` reference, if I provided the values, txid, and index, I think.

kallewoof replied:
Actually, I can probably use `CInputCoin` if I keep the value ivar. Will try that.

kallewoof replied:
`CInputCoin` worked great after some tweakery.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -3112,6 +3119,30 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
+        tmp_cc.m_avoid_partial_spends = true;
+        CAmount nFeeRet2;
+        int nChangePosInOut2 = nChangePosIn;
+        if (_create_tx(vecSend, tx2, reservekey, nFeeRet2, nChangePosInOut2, strFailReason, tmp_cc, sign)) {
```
Can this branching for avoid_partial_spends on/off be done in `CWallet::SelectCoins` (where all different combinations are tried), instead of creating a second place of iterating through things?

kallewoof replied:
I would love to, but it relies on absolute fee to determine whether to force-use the avoid partial variant, which is not available in `SelectCoins`.

sipa commented on src/wallet/coinselection.cpp:
```diff
@@ -244,7 +244,10 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<OutputGroup>& group
     }
 
-    if (nTotalLower < nTargetValue) {
+    // we prioritize the lowest larger if
```
This feels like a completely unrelated change to the coin selection algorithm.

kallewoof replied:
Dropped the relevant commit. I will make a PR once I have simulation results to back the change after merge.

sipa reviewed (commented):
Overall: can you follow code style conventions for method names (CamelCase). In addition and relatedly, having methods named `push_back` on application-level classes is very confusing as it makes it look like they're STL containers when they're not even behaving like the equivalently named STL method.

MarcoFalke commented on src/wallet/coincontrol.cpp:
```diff
@@ -0,0 +1,23 @@
+    fOverrideFeeRate = false;
+    m_confirm_target.reset();
+    m_signal_bip125_rbf.reset();
+    m_fee_mode = FeeEstimateMode::UNSET;
```
Generally, we initialize all non-static class members where they are declared. Ensure determinism by avoiding accidental use of uninitialized values. Also, static analyzers balk about this. Initializing the members in the declaration makes it easy to spot uninitialized ones.

https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#c-data-structures

kallewoof replied:
It would duplicate the initialization in the `SetNull()` method as well as in the variable declarations. I assume that's okay, then?

MarcoFalke replied:
Ah, sorry. I had missed that `SetNull()` is used in the gui code once.

kallewoof replied:
NP! I am unsure if c942091 should be dropped or if it's a-good to have. Keeping it unless/until someone complains.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -669,6 +668,39 @@ struct CoinEligibilityFilter
+    , m_descendants(descendants)
+    , effective_value(0)
+    , fee(0)
+    , long_term_fee(0)
```
nit: same for all members here. Can just inline into `CAmount long_term_fee{0};` above to avoid duplication.

kallewoof committed:
utils: Add insert() convenience templates

kallewoof committed:
Add -avoidpartialspends and m_avoid_partial_spends

kallewoof committed:
wallet: Add input bytes to CInputCoin

With nInputBytes, coin selection can execute without a reference to the COutput

kallewoof committed:
clean-up: Remove no longer used ivars from CInputCoin

kallewoof commented:
Rebased and addressed all nits except [@sipa's request to move the 'try avoiding' part into SelectCoins](https://github.com/bitcoin/bitcoin/pull/12257#discussion_r202492813): I cannot access fees easily, since select coins handles sets of coins, rather than transactions. It may be reasonable to switch to simply using 'number of inputs' rather than 'absolute fee', in which case this can be moved to `SelectCoins`

kallewoof commented:
(mostly @sipa) The policy change in 78f75da is actually a necessary part of the algorithm, so it was re-added. Consider sending 1.1 btc from this utxo set:
* 1.0 btc to A
* 0.3 btc to A
* 0.2 btc to B
* 1.0 btc to C

The implementation change will see three "coins": A (1.3 btc), B (0.2), and C (1.0). It will see A as 'lowest larger', and not use it, because sum(B:C) can cover the payment.

Ideally, we want it to pick A, unless A is really big and/or our payment is really small (I arbitrarily set this limit to 2*target).

sipa commented:
@kallewoof I understand your argument that it may be desirable, but I don't understand why it is necessary.

How is it different from a scenario in the existing code where there are only 3 UTXOs (with sizes 1.3 BTC, 0.2 BTC, and 1 BTC)?

kallewoof commented:
After discussing with @sipa (and after realizing this only applies to the knapsack solver which is intended to be removed in the near future) I removed the policy change again.

kallewoof committed:
doc: Add release notes for -avoidpartialspends

gmaxwell commented:
ACK

sdaftuar commented on src/wallet/coinselection.cpp:
```diff
@@ -298,3 +298,35 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<CInputCoin>& vCoins
+    m_value += output.effective_value;
+    m_depth = std::min(m_depth, depth);
+    m_ancestors = std::max(m_ancestors, ancestors);
+    m_descendants = std::max(m_descendants, descendants);
```
Can you explain why this is taking `std::max`?  I think the correct calculation would involve taking the set of ancestors or descendants of the group and counting the unique entries.

kallewoof replied:
The `descendants` value is already a complete count of all descendants starting from the top ancestor of the transaction in question. Here we only really care about `who has the most`.

sdaftuar replied:
I believe this calculation may be fine for descendants (if I'm correctly understanding what the descendants variable represents -- it could use a code comment explaining exactly what it is, I think the `descendants` variable is the max number of descendants of any ancestor of a given coin?), however I think it is incorrect for ancestors.

I believe the ancestor count is the actual number of ancestors a given coin has.  If you just take the max across all coins in an output group, then if you have e.g. 30 2-ancestor coins in an output group, you may treat them as a 2-ancestor output group, when in fact you have 60 ancestors in any transaction that includes the group.

(For descendants, I think this works out fine, because all we're doing by creating a new transaction is adding one more descendant to every ancestor, so considering the max value is sufficient to try to avoid hitting mempool chain limits.)

I understand that the existing logic may be imprecise and hacky around this, but I believe that combining all the coins into an output group exacerbates this issue.

Regardless of what we end up doing here though I think the code could use a comment explaining the logic, as this is non-obvious.

kallewoof replied:
> I believe the ancestor count is the actual number of ancestors a given coin has. If you just take the max across all coins in an output group, then if you have e.g. 30 2-ancestor coins in an output group, you may treat them as a 2-ancestor output group, when in fact you have 60 ancestors in any transaction that includes the group.

Aha! This is a very good point. I will think about a remedy for this post-merge. For now, I added a comment explaining how it works and how it can be improved.

sipa commented on src/wallet/wallet.h:
```diff
@@ -642,16 +642,7 @@ struct CoinSelectionParams
-    CoinEligibilityFilter(int conf_mine, int conf_theirs, uint64_t max_ancestors) : conf_mine(conf_mine), conf_theirs(conf_theirs), max_ancestors(max_ancestors), max_descendants(max_ancestors) {}
-    CoinEligibilityFilter(int conf_mine, int conf_theirs, uint64_t max_ancestors, uint64_t max_descendants) : conf_mine(conf_mine), conf_theirs(conf_theirs), max_ancestors(max_ancestors), max_descendants(max_descendants) {}
-};
+struct CoinEligibilityFilter;
```
No need for a forward declaration here. wallet/wallet.h already includes wallet/coinselection.h where it is defined.

sipa commented on src/wallet/coincontrol.h:
```diff
@@ -17,23 +17,23 @@ class CCoinControl
 public:
     //! Custom change destination, if not set an address is generated
-    CTxDestination destChange;
+    CTxDestination destChange{CNoDestination()};
```
If all local variables have a default initialized, you don't need the SetNull call in the constructor anymore.

kallewoof replied:
They don't -- specifically the new argument requires `gArgs` which are not available here.

I should probably drop "Declare-&-initialize..." commit.

kallewoof replied:
Dropped that commit.

sipa commented on src/wallet/coinselection.h:
```diff
@@ -66,8 +66,38 @@ struct CoinEligibilityFilter
+
+    OutputGroup() {}
+    OutputGroup(std::vector<CInputCoin>&& outputs, bool from_me, CAmount value, int depth, size_t ancestors, size_t descendants)
+    : m_outputs(outputs)
```
More efficient: `m_outputs(std::move(outputs))`. The expression 'outputs' here is not an rvalue reference, so this is not automatic.

kallewoof replied:
Huh.. I thought `&&` made it an rvalue. Guess I need to re-read that explanation. Done.

sipa commented on src/wallet/coinselection.h:
```diff
@@ -9,6 +9,8 @@
 #include <primitives/transaction.h>
 #include <random.h>
 
+struct OutputGroup;
```
What is the forward declare for?

kallewoof replied:
Leftover. Removed.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -2635,8 +2641,8 @@ OutputType CWallet::TransactionChangeType(OutputType change_type, const std::vec
 
-bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CReserveKey& reservekey, CAmount& nFeeRet,
-                                int& nChangePosInOut, std::string& strFailReason, const CCoinControl& coin_control, bool sign)
+bool CWallet::_create_tx(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CReserveKey& reservekey, CAmount& nFeeRet,
```
That's a really crazy function name. What about `CreateTransactionInternal` or so?

kallewoof replied:
Renamed.

kallewoof committed:
moveonly: CoinElegibilityFilter into coinselection.h

kallewoof committed:
wallet: Add output grouping

kallewoof committed:
wallet: Switch to using output groups instead of coins in coin selection

sdaftuar commented on src/wallet/wallet.cpp:
```diff
@@ -3022,6 +3011,30 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
+    int nChangePosIn = nChangePosInOut;
+    CTransactionRef tx2 = tx;
+    bool res = CreateTransactionInternal(vecSend, tx, reservekey, nFeeRet, nChangePosInOut, strFailReason, coin_control, sign);
+    // try with avoidpartialspends unless it's enabled already
```
I think I would prefer that we not do this -- I'm not sure that `-avoidpartialspends` works very well on wallets that have substantial address reuse, in particular I think you can devolve into cases where you'd produce giant transactions that take forever to sign and would never pass policy (or consensus) limits, in extreme cases.

I'm going to comment elsewhere about places I think we can mitigate this sort of pathological behavior, but as belt and suspenders I think if the user is not using this new feature, we should not fall back to trying it anyway.

kallewoof replied:
The code only selects the avoiding tx if its fee is lower than the original. I don't think you'd ever run into that case for someone with substantial address reuse.

sdaftuar replied:
But even running the code on a wallet with a large number of outputs with the same scriptPubKey can be very slow.  Nothing stops the code from eg trying to generate a multi-megabyte transaction and then hit a huge slow down while trying to sign such a transaction, before the transaction gets rejected.  (I have an RPC test that demonstrates this possibility.)

kallewoof replied:
Now that there is a cap on the output group count, I think parts of this is mitigated, but I'm not sure I've convinced myself this is necessary. Perhaps a 3-state flag (try, on, off) with default=try, so people can disable it...

kallewoof replied:
Removing this. No harm in doing the try style later.

sdaftuar commented on src/wallet/wallet.cpp:
```diff
@@ -4379,3 +4392,25 @@ void CWallet::LearnAllRelatedScripts(const CPubKey& key)
     LearnRelatedScripts(key, OutputType::P2SH_SEGWIT);
 }
 
+std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outputs, bool single_coin) const {
```
Perhaps we should not create groups that would be excessively large to send together?  For instance if someone has a couple thousand outputs to the same address, then as a group they would become unspendable (because the resulting transaction would be too big).  

It might be pretty annoying for our users if they became vulnerable to weird DoS attacks where people create lots of dust to prevent them from spending their coins...

kallewoof replied:
Good point. I've set a cap at 10 outputs per group.

That being said, the code (at least BnB algo; knapsack is being deprecated) already removes dust from the group before running coin select. https://github.com/kallewoof/bitcoin/blob/19bbab013860c84475cce98a55179a1167348269/src/wallet/wallet.cpp#L2415-L2422

laanwj commented on src/util.h:
```diff
@@ -355,4 +355,14 @@ std::string CopyrightHolders(const std::string& strPrefix);
 
+//! Simplification of std insertion
+template <typename Tdst, typename Tsrc>
+void insert(Tdst& dst, const Tsrc& src) {
```
should we call these `insert`? I think it's a bit confusing for readability, as it implies the built-in std function is used in ways that the original one cannot be.

kallewoof replied:
I was considering putting them under a bitcoin:: namespace or somehing to emphasize they're actually not `std` but since we are no longer `using namespace std` it felt unnecessary. I'm up for renaming if it still feels confusing though.

kallewoof replied:
I ended up moving into `util::` after all.

laanwj commented on src/wallet/coinselection.h:
```diff
@@ -28,12 +28,20 @@ class CInputCoin {
     CAmount long_term_fee = 0;
 
+    /** Pre-computed estimated size of this output as a fully-signed input in a transaction. Can be -1 if it could not be calculated */
+    int m_input_bytes{-1};
```
what makes this different from
```
int m_input_bytes = -1;
```
(my guess is that it is the same, in which case I think that syntax is preferable because it is already used in so many places)

kallewoof replied:
It is the same. We already use this syntax in a variety of places. See e.g. 

https://github.com/bitcoin/bitcoin/blob/7ebd8c6385253bfa30175169a22f93474d8ddfaf/src/wallet/wallet.h#L97-L101

https://github.com/bitcoin/bitcoin/blob/7ebd8c6385253bfa30175169a22f93474d8ddfaf/src/wallet/wallet.h#L674-L675

laanwj replied:
okay! well then my only comment is that mixing the syntaxes in the same class looks inconsistent, but I don't care strongly

kallewoof replied:
Oh, hum. I actually thought the {} syntax was preferred and didn't fix the other stuff to avoid diff bloat.

kallewoof replied:
Oh, actually, the other two lines are removed at the end as they are no longer needed.

laanwj commented:
utACK 232f96f5c8a3920c09db92f4dbac2ad7d10ce8cf

sdaftuar reviewed (commented):
Oops, I was mid-review and never posted these comments before merge.  I address one of these in #13805.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4438,3 +4427,29 @@ void CWallet::LearnAllRelatedScripts(const CPubKey& key)
+            size_t ancestors, descendants;
+            mempool.GetTransactionAncestry(output.tx->GetHash(), ancestors, descendants);
+            if (!single_coin && ExtractDestination(output.tx->tx->vout[output.i].scriptPubKey, dst)) {
+                if (gmap.count(dst) == 10) {
```
I'm not sure it's a big risk, but I also think it's better to be safe than sorry. I envision a replacement of `push_back` with `insert(begin() + random(size())` (forgive pseudo-code-abbrev) in

https://github.com/bitcoin/bitcoin/blob/4d550ffab64d79a44fcaa76f6e30c28f01614d3f/src/wallet/coinselection.cpp#L298

Will make a PR unless someone beats me to it.

kallewoof replied:
Hm.. actually, that's not sufficient due to the capping at 10. I guess shuffling the entire vector at the top in `CWallet::GroupOutputs` is the way to go.

kallewoof replied:
Adding shuffling in #13808. I chose to put it in the `GroupOutputs` method since it can skip the shuffle for `is_single` and `size() <= 10` cases.

promag replied:
`gmap.count(dst) == 10` is always false and is not fixed in #13808.

kallewoof replied:
It's fixed in #13805

kallewoof commented on src/wallet/coinselection.cpp:
```diff
@@ -298,3 +287,40 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<CInputCoin>& vCoins
+    // m_ancestors is currently the max ancestor count for all coins in the group; however, this is
+    // not ideal, as a wallet will consider e.g. thirty 2-ancestor coins as having two ancestors,
+    // when in reality it has 60 ancestors.
+    m_ancestors = std::max(m_ancestors, ancestors);
```
Yeah, worst case the user has to wait for the inputs to confirm before they can spend them. (Is there a case where that could be a pontential problem? Can't think of one.)

kallewoof replied:
See #13812
